### PR TITLE
Logger

### PR DESCRIPTION
**Pull request by Tronic**
A bigger reworking of logger. NOT READY FOR MERGE but here for comments. As can be seen from testing messages on exit, it is still not thread-safe (messages get repeated). Perhaps we need to give up using std::clog in order to fix it -- I honestly don't know at this time -- but for whatever reason random repeats seem to occur within cerr and/or boost streambuf whenever multiple threads write to clog at the same time, and I don't think this can be fixed by the mutex in log.cc.

### CODE REVIEW

**Comment by Tronic** [comment]
The beef is really in log.cc (everything else is about fixing log messages). That was quite an interesting exercise at mixing POSIX and modern C++.

File: game/audio.cc [hunk 1]
```diff
@@ -423,7 +423,12 @@ struct Audio::Impl {
 	std::string selectedBackend = Audio::backendConfig().getValue();
 	Impl() {
 		populateBackends(portaudio::AudioBackends().getBackends());
-		std::clog << portaudio::AudioBackends().dump() << std::flush; // Dump PortAudio backends and devices to log.
+		{
```
Comment (Baklap4) [inline game/audio.cc]: Why the extra braces?

File: game/log.cc [hunk 2]
```diff
@@ -37,159 +38,172 @@
...
-std::string target;
-int minLevel;
-
-void writeLog(std::string const& msg) {
-	std::lock_guard<std::mutex> l(log_lock);
-	std::cerr << msg << std::flush;
-	file << msg << std::flush;
-}
+static struct Impl {
+	const std::map<std::string, int> numeric = {
```
Comment (Baklap4) [inline game/log.cc]: Personally i'd rename this to `loglevels` or something like that. Numeric is pretty much not saying anything

File: game/audio.cc (hunk 1 above)
Comment (Tronic) [reply game/audio.cc]: To limit the lifetime of logmsg to that block.

**Comment by Tronic** [comment]
Although this branch has relevant improvements, I doubt that anyone will have time to finish the work. or even bring it up to date. Closing.

**Pull request closed without merging**